File: src/main/java/org/gangel/graphomance/neo4j/NeoConnectionProducer.java
package org.gangel.graphomance.neo4j;

import org.gangel.graphomance.*;
import org.neo4j.graphdb.GraphDatabaseService;
import org.neo4j.graphdb.factory.GraphDatabaseFactory;

import java.io.File;
import java.util.Objects;

public class NeoConnectionProducer implements ConnectionProducer, SessionProducer {

    @Override
    public Connection connect(ConnectionSettings settings) {
        NeoConnectionSettings opts = Objects.requireNonNull((NeoConnectionSettings) settings, "Requiring Neo4j connection settings");

        GraphDatabaseService dbService = new GraphDatabaseFactory().newEmbeddedDatabase(new File(opts.getDbPath()));
        registerShutdownHook( dbService );

        return NeoConnection.builder().dbService(dbService).build();
    }

    private static void registerShutdownHook( final GraphDatabaseService graphDb )
    {
        // Registers a shutdown hook for the Neo4j instance so that it
        // shuts down nicely when the VM exits (even if you "Ctrl-C" the
        // running application).
        Runtime.getRuntime().addShutdownHook( new Thread( () -> graphDb.shutdown() ));
    }

    @Override
    public Session createSession(Connection connection) {
        return new NeoSession(Objects.requireNonNull((NeoConnection) connection, "Requiring Neo4j connection object"));
    }
}

File: src/main/kotlin/org/gangel/graphomance/neo4j/NeoConnectionProducer.kt
package org.gangel.graphomance.neo4j

import org.gangel.graphomance.*
import org.neo4j.driver.AuthToken
import org.neo4j.driver.AuthTokens
import org.neo4j.driver.GraphDatabase

class NeoConnectionProducer : ConnectionProducer, SessionProducer {
	override fun connect(settings: ConnectionSettings): Connection {
		if (settings !is NeoConnectionSettings)
			throw RuntimeException("Requiring Neo4j connection settings")
		val opts: NeoConnectionSettings = settings
		return NeoConnection(GraphDatabase.driver(opts.dbPath, AuthTokens.basic("neo4j", "password")))
	}

	override fun createSession(connection: Connection): Session {
		val neoConnection = connection as? NeoConnection ?: throw RuntimeException("Requiring Neo4j connection object")
		return NeoSession(neoConnection)
	}

}